### PR TITLE
Remove dependency from `LogicalPlan::TableScan` to `ExecutionPlan`

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2247

 # Rationale for this change

We want to remove the dependency from `TableScan` to `ExecutionPlan` so that we can move the `LogicalPlan` into the `expr` crate.

# What changes are included in this PR?

- New `TableSource` trait in `expr` crate that is referenced from `TableScan`
- New `DefaultTableSource` struct in core crate that wraps `TableProvider`
- Shim methods

# Are there any user-facing changes?

Yes, at API level

### CODE REVIEW

**Comment by andygrove** [comment]
@matthewmturner @alamb I aborted the large refactor and found a simpler and more practical approach. Let me know what you think.

**Comment by alamb** [comment]
I plan to review this later today

**Comment by alamb** [comment]
The last comment from @Jimexist during "the great split of datafusion" was is https://github.com/apache/arrow-datafusion/pull/1762#discussion_r802563462 --  I think this PR will help

File: datafusion/core/src/logical_plan/plan.rs [hunk 1]
```diff
@@ -125,13 +127,71 @@ pub struct Window {
     pub schema: DFSchemaRef,
 }
 
+/// DataFusion default table source, wrapping TableProvider
```
Comment (alamb) [inline datafusion/core/src/logical_plan/plan.rs:130]:
```suggestion
/// DataFusion default table source, wrapping TableProvider
///
/// This structure adapts a `TableProvider` (physical plan trait) to the `TableSource` 
/// (logical plan trait)
```

File: datafusion/core/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -125,13 +127,71 @@ pub struct Window {
...
+
+/// Wrap TableProvider in TableSource
+pub fn provider_as_source(
+    table_provider: Arc<dyn TableProvider>,
+) -> Arc<dyn TableSource> {
+    Arc::new(DefaultTableSource::new(table_provider))
+}
+
+/// Extract TableProvider from TableSource
+pub fn source_as_provider(
```
Comment (alamb) [inline datafusion/core/src/logical_plan/plan.rs:173]: Maybe it is worth a comment on `TableSource` that it is not intended, initially, to be extended outside of DataFusion.

File: datafusion/expr/src/table_source.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+pub enum TableType {
+    /// An ordinary physical table.
+    Base,
+    /// A non-materialised table that itself uses a query internally to provide data.
+    View,
+    /// A transient table.
+    Temporary,
+}
+
+/// TableSource
```
Comment (alamb) [inline datafusion/expr/src/table_source.rs:52]: It is probably worth some comments here about the rationale for this structure existing -- namely so LogicalPlan doesn't directly rely on physical plans

**Review by alamb — APPROVED** [review]
Looks good to me -- thanks @andygrove 

cc @xudong963 @yjshen

**Comment by andygrove** [comment]
Thanks for the review @alamb. I pushed an update with more documentation. Let me know if I missed anything and I can address in one of the following refactor PRs.

**Pull request merged**